### PR TITLE
Create App

### PR DESCRIPTION
**Pull request by sithumonline**
image
---
image
---
image
---
Known Issues :: CLI gives type error when the environment wasn't set
```    
TypeError: Cannot read property 'ID' of undefined
```

### CODE REVIEW

File: src/commands/app/env.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      }
+      ,
+      environment: {
+        name: flags.environment,
+        required: true
+      }
+    });
+    const ctx = context as Required<typeof context>;
+    const { orgId, projectId, envId } = ctx;
+    cli.action.stop('\nConfigured environment for app');
```
Comment (ChathurangaKCD) [inline src/commands/app/env.ts:104]: Should be `cli.action.start` , also need to change the message.

File: src/commands/app/con.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      throw new Error('Please Set App Environment');
+    }
+    const data: AppCreateContainer = {
+      ID: app.ID,
+      name: flags.containerName,
+      orgId: orgId,
+      projectId: projectId,
+      appEnvId: appEnvId,
+      type: flags.appType,
+      cpu: flags.cpu,
```
Comment (ChathurangaKCD) [inline src/commands/app/con.ts:112]: should be validated for min/max

File: src/commands/app/con.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      memory: flags.memory,
+      port: flags.port
```
Comment (ChathurangaKCD) [inline src/commands/app/con.ts:114]: multiple ports?, also needs to be disabled for `INIT` type

File: src/commands/app/env.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }),
+    appName: flags.string({
+      char: 'n',
+      description: 'App Name',
+      required: true,
+      multiple: false
+    }),
+    appType: flags.string({
+      char: 't',
+      description: 'Service Type (ClusterIP|NodePort|LoadBalancer)',
```
Comment (ChathurangaKCD) [inline src/commands/app/env.ts:38]: need to validate enum

**Review by ChathurangaKCD — CHANGES_REQUESTED** [review]

**Review by Shehanka — APPROVED** [review]

File: src/commands/app/env.ts [hunk 5]
```diff
@@ -85,31 +65,16 @@ export default class Init extends Command {
...
     if (!flags.appType.match(/^(ClusterIP|NodePort|LoadBalancer)$/)) {
       throw new Error('Wrong app type, it must be ClusterIP,NodePort or LoadBalancer');
     }
-    let appName: string | undefined = flags.appName;
-    if (!flags.appName) {
-      appName = getDefaultAppName();
-    }
+    const projectId = getDefaultProjectIdFile();
+    const orgId = getDefaultOrganizationIdFile();
+    const appName = getDefaultAppName();
```
Comment (ChathurangaKCD) [inline src/commands/app/env.ts:74]: reads the file 3 times.
define an interface for file data, & read once.

File: src/modules/apps/files.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  try {
+    if (!fs.existsSync(appFile)) {
+      throw new Error(appFile + ' not found, please init app first.');
+    }
+    data = fs.readFileSync(appFile, 'utf-8');
+  } catch (e) {
+    throw e;
+  }
+  try {
+    json = JSON.parse(data);
```
Comment (ChathurangaKCD) [inline src/modules/apps/files.ts:27]:
```suggestion
    return JSON.parse(data);
```

File: src/modules/util/validations.ts [hunk 7]
```diff
@@ -82,6 +82,17 @@ export async function tryValidateCommonFlags(flags: {
   return {
     orgId,
     projectId,
-    envId,
+    envId
   };
 }
+
+export async function ValidateEnvironment(organizationId: string, projectId: string, environmentName: string) {
+  let envId;
```
Comment (ChathurangaKCD) [inline src/modules/util/validations.ts:90]:
```suggestion
  const envId = (await getEnvironmentIdByName(organizationId, projectId, environmentName)).
    ?.ID;
```

File: src/commands/app/env.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    const ctx = context as Required<typeof context>;
+    const { envId } = ctx;
+    const app = await getApp({ projectId: fileData.projectId, orgId: fileData.orgId, name: fileData.name });
+    if (!app) {
+      throw new Error('App not found');
+    }
+    if (app.app_environments?.some(a => a.environment_id === envId)) {
+      throw new ValidationError('App environment already initialized');
+    }
+    await ensureTargetNamespace({ orgId: fileData.orgId, projectId: fileData.projectId, envId, name: flags.namespace });
```
Comment (ChathurangaKCD) [inline src/commands/app/env.ts:83]: shouldn't create a namespace, if it doesn't exist

File: src/modules/apps/files.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    throw e;
+  }
+  try {
+    return JSON.parse(data);
+  } catch (e) {
+    throw new Error('Failed to parse json');
+  }
+}
+
+export function writeFile(content: any) {
```
Comment (ChathurangaKCD) [inline src/modules/apps/files.ts:33]:
```suggestion
export function writeFile(content: AppCreate) {
```

**Review by ChathurangaKCD — CHANGES_REQUESTED** [review]

**Pull request merged**